src/psyclone/psyir/nodes/routine.py

```diff
@@ -64,19 +73,29 @@ class Routine(Schedule, CommentableMixin):
     _children_valid_format = "[Statement]*"
     _text_name = "Routine"
 
-    def __init__(self, name, is_program=False, **kwargs):
+    def __init__(self, name, is_program=False, symbol=None,
+                 symbol_tag=None, **kwargs):
```

Comment:
If this accepts pre-existing symbols, I would prefer this to not deal with tags. I think doing the symbol in a separate step is reasonable for this cases, but this all follows from the assumption that pre-existing symbols are fine. Is this the case?

Reply:
Ok - so the assumption at the moment is that pre-existing symbols are fine, but they're not expected to be in the parent's symbol table already (since the assumption is the Routine is meant to be in control of that).

Part of the reason the Routine needs to know about the symbol_tag is to ensure its carried over if the symbol is detached and then reattached, e.g. during a tree copy. I'm not sure if the tags are required for any correctness in the `FortranWriter`, but if someone wanted to copy a tree and do operations on it, the tags would need to be copied and the only way this can happen at the moment is if the Routine knows about the tag itself.

Does this make sense @sergisiso

Reply:
>  Part of the reason the Routine needs to know about the symbol_tag is to ensure its carried over if the symbol is detached and then reattached, e.g. during a tree copy

A tree copy should do a symbol_table.deep_copy for every scope, this will copy the symbols and add the same tags to the newly equivalent instances of the symbols.

This implementation had several issues that @arporter fixed in a PR that we merged yesterday, could you bring this branch to master and see if it is still a problem?

Reply:
Yes - but this doesn't work with this new Routine implementation (see Container._refine_copy) - in fact this is the main challenge here.

Since copy makes a copy of the tree, it will make a copy of the Routines and a copy of the symbols but they're not linked to each other. The _refine_copy removes the copied symbols to be replaced by the Routine parent setter so the tag is lost unless the Routine has it - fixing the `deep_copy` in symbol table won't fix this issue.